First MVC: Controller and View tutorials: MT sweep all issues fixed (#21379)
dotnet/AspNetCore.Docs: aspnetcore/tutorials/first-mvc-app/adding-view.md

::: moniker range=">= aspnetcore-3.0"

In this section, you:

* Modify the `HelloWorldController` class to use [Razor](xref:mvc/views/razor) view files. This cleanly encapsulates the process of generating HTML responses to a client.
* Create a view template file using Razor. Razor-based view templates:

> 

Reviewer: Delete this. No reason for blank like.

  * Have a *.cshtml* file extension.
  * Provide an elegant way to create HTML output with C#.

Currently the `Index` method returns a string with a message that's hard-coded in the controller class. In the `HelloWorldController` class, replace the `Index` method with the following code:

[!code-csharp[](~/tutorials/first-mvc-app/start-mvc/sample/MvcMovie/Controllers/HelloWorldController.cs?name=snippet_4)]

The preceding code:

* Calls the controller's <xref:Microsoft.AspNetCore.Mvc.Controller.View*> method.
* Uses a view template to generate an HTML response.

Controller methods:

> * Are also known as *action methods*, such as the `Index` method in the preceding code.

Reviewer: ```suggestion
* Are referred to as  *action methods*.  For example, the `Index` action method in the preceding code.
```

> * Generally return an <xref:Microsoft.AspNetCore.Mvc.IActionResult> (or a class derived from <xref:Microsoft.AspNetCore.Mvc.ActionResult>), not a type like `string`.

Reviewer: ```suggestion
* Generally return an <xref:Microsoft.AspNetCore.Mvc.IActionResult> or a class derived from <xref:Microsoft.AspNetCore.Mvc.ActionResult>, not a type like `string`.
```

# [Visual Studio Code](#tab/visual-studio-code)

Add an `Index` view for the `HelloWorldController`:

* Add a new folder named *Views/HelloWorld*.
* Add a new file to the *Views/HelloWorld* folder name *Index.cshtml*.

# [Visual Studio for Mac](#tab/visual-studio-mac)

Control-click on the *Views* folder, and then **Add > New Folder** and name the folder *HelloWorld*.

Control-click on the *Views/HelloWorld* folder, and then **Add > New File**.

In the **New File** dialog:

* Select **ASP.NET Core** in the left pane.
* Select **Razor View - Empty** in the center pane.
* Type *Index* in the **Name** box.
* Select **New**.

  ![Add New Item dialog](adding-view/_static/add_view_macVSM8.9.png)

---

Replace the contents of the *Views/HelloWorld/Index.cshtml* Razor view file with the following:

[!code-cshtml[](~/tutorials/first-mvc-app/start-mvc/sample/MvcMovie22/Views/HelloWorld/Index1.cshtml?highlight=7)]

Navigate to `https://localhost:{PORT}/HelloWorld`:

* The `Index` method in the `HelloWorldController` ran the statement `return View();`, which specified that the method should use a view template file to render a response to the browser.

> * A view template file name wasn't specified, MVC defaulted to using the default view file. The default view file has the same name as the method (`Index`), so the view template in */Views/HelloWorld/Index.cshtml* is used.

Reviewer: ```suggestion
* A view template file name wasn't specified, so MVC defaulted to using the default view file. When the view file name isn't specified, the default view is returned. The default view has the same name as the action method, `Index` in this example. The view template */Views/HelloWorld/Index.cshtml* is used.
```

> * The following image shows the string "Hello from our View Template!" hard-coded in the view:

Reviewer: hard-coded is slang and doesn't MT. It's not needed here, they see where it comes from. This only makes sense once you explain dynamic values.

  ![Browser window](~/tutorials/first-mvc-app/adding-view/_static/hell_template.png)

## Change views and layout pages

> Select the menu links: **MvcMovie**, **Home**, and **Privacy**. Each page shows the same menu layout. The menu layout is implemented in the *Views/Shared/_Layout.cshtml* file.

Reviewer: ```suggestion
Select the menu links **MvcMovie**, **Home**, and **Privacy**. Each page shows the same menu layout. The menu layout is implemented in the *Views/Shared/_Layout.cshtml* file.
```

Open the *Views/Shared/_Layout.cshtml* file.

> [Layout](xref:mvc/views/layout) templates allow you to:
> 
> * Specify the HTML container layout of a site in one place.
> * Apply the HTML container layout across multiple pages in the site.

Reviewer: ```suggestion
[Layout](xref:mvc/views/layout) templates allows:

* Specifying the HTML container layout of a site in one place.
* Applying the HTML container layout across multiple pages in the site.
```

The *Views/_ViewStart.cshtml* file brings in the *Views/Shared/_Layout.cshtml* file to each view. The `Layout` property can be used to set a different layout view, or set it to `null` so no layout file will be used.

Open the *Views/HelloWorld/Index.cshtml* view file.

Change the title and `<h2>` element as highlighted in the following:

[!code-cshtml[](~/tutorials/first-mvc-app/start-mvc/sample/MvcMovie/Views/HelloWorld/Index2.cshtml?highlight=2,5)]

> The title and `<h2>` element are slightly different so it's clear which bit of code changes the display.

Reviewer: ```suggestion
The title and `<h2>` element are slightly different so it's clear which part of the code changes the display.
```

![Movie List view](~/tutorials/first-mvc-app/adding-view/_static/hell3.png)
![Movie List view](~/tutorials/first-mvc-app/adding-view/_static/hell50.png)

Our little bit of "data" (in this case the "Hello from our View Template!" message) is hard-coded, though. The MVC application has a "V" (view) and you've got a "C" (controller), but no "M" (model) yet.
The small bit of "data", the "Hello from our View Template!" message, is hard-coded however. The MVC application has a "V" (view), a "C" (controller), but no "M" (model) yet.

## Passing Data from the Controller to the View

Controller actions are invoked in response to an incoming URL request. A controller class is where the code is written that handles the incoming browser requests. The controller retrieves data from a data source and decides what type of response to send back to the browser. View templates can be used from a controller to generate and format an HTML response to the browser.

> Controllers are responsible for providing the data required in order for a view template to render a response. A best practice: View templates should **not** perform business logic or interact with a database directly. Rather, a view template should work only with the data that's provided to it by the controller. Maintaining this "separation of concerns" helps keep the code clean, testable, and maintainable.

Reviewer: Prefer the original. Sentences are not long.

> View templates should **not**:

Reviewer: remove bullet list and restore original sentence

A view template should work only with the data that's provided to it by the controller. Maintaining this "separation of concerns" helps keep the code:

* Clean.
* Testable.
* Maintainable.

Currently, the `Welcome` method in the `HelloWorldController` class:

* Takes a `name` and a `ID` parameter.
* Outputs the values directly to the browser.

> Rather than have the controller render this response as a string, change the controller to use a view template instead. The view template generates a dynamic response, which means that appropriate bits of data must be passed from the controller to the view to generate the response. Do this by having the controller put the dynamic data (parameters) that the view template needs, in a `ViewData` dictionary. The view template can then access the dynamic data.

Reviewer: ```suggestion
Rather than have the controller render this response as a string, change the controller to use a view template instead. The view template generates a dynamic response, which means that appropriate data must be passed from the controller to the view to generate the response. Do this by having the controller put the dynamic data (parameters) that the view template needs in a `ViewData` dictionary. The view template can then access the dynamic data.
```

In *HelloWorldController.cs*, change the `Welcome` method to add a `Message` and `NumTimes` value to the `ViewData` dictionary.

> The `ViewData` dictionary is a dynamic object, which means any type can be used; the `ViewData` object has no defined properties until you put something inside it. The [MVC model binding system](xref:mvc/models/model-binding) automatically maps the named parameters (`name` and `numTimes`) from the query string in the address bar to parameters in your method. The complete *HelloWorldController.cs* file looks like this:

Reviewer: ```suggestion
The `ViewData` dictionary is a dynamic object, which means any type can be used. The `ViewData` object has no defined properties until something is added. The [MVC model binding system](xref:mvc/models/model-binding) automatically maps the named parameters `name` and `numTimes` from the query string to parameters in the method. The complete `HelloWorldController`:
```

`https://localhost:{PORT}/HelloWorld/Welcome?name=Rick&numtimes=4`

Data is taken from the URL and passed to the controller using the [MVC model binder](xref:mvc/models/model-binding). The controller packages the data into a `ViewData` dictionary and passes that object to the view. The view then renders the data as HTML to the browser.

![Privacy view showing a Welcome label and the phrase Hello Rick shown four times](~/tutorials/first-mvc-app/adding-view/_static/rick2_50.png)

> In the sample above, the `ViewData` dictionary was used to pass data from the controller to a view. Later in the tutorial, a view model is used to pass data from a controller to a view. The view model approach to passing data is preferred over the `ViewData` dictionary approach. For more information, see [When to use ViewBag, ViewData, or TempData](https://www.rachelappel.com/when-to-use-viewbag-viewdata-or-tempdata-in-asp-net-mvc-3-applications/).

Reviewer: ```suggestion
In the preceding sample, the `ViewData` dictionary was used to pass data from the controller to a view. Later in the tutorial, a view model is used to pass data from a controller to a view. The view model approach to passing data is preferred over the `ViewData` dictionary approach.
```

In the next tutorial, a database of movies is created.

> [!div class="step-by-step"]
> [Previous: Add a Controller](adding-controller.md)
> [Next: Add a Model](adding-model.md)

::: moniker-end

> ::: moniker range="< aspnetcore-3.0"

Reviewer: Don't modify previous versions.

In this section, the `HelloWorldController` class is modified to use [Razor](xref:mvc/views/razor) view files to cleanly encapsulate the process of generating HTML responses to a client.

> You create a view template file using Razor. Razor-based view templates have a *.cshtml* file extension. They provide an elegant way to create HTML output with C#.

Reviewer: ```suggestion
View templates are created using Razor. Razor-based view templates have a *.cshtml* file extension. They provide an elegant way to create HTML output with C#.
```

Author: Resolved as you suggested in the 5.x version (line 20).  I did not change the older version specified here on line 241.

Currently the `Index` method returns a string with a message that's hard-coded in the controller class. In the `HelloWorldController` class, replace the `Index` method with the following code:

[!code-csharp[](~/tutorials/first-mvc-app/start-mvc/sample/MvcMovie/Controllers/HelloWorldController.cs?name=snippet_4)]

The preceding code:

> * Calls the controller's <xref:Microsoft.AspNetCore.Mvc.Controller.View*> method. It uses a view template to generate an HTML response. Controller methods (also known as *action methods*), such as the `Index` method above, generally return an <xref:Microsoft.AspNetCore.Mvc.IActionResult> (or a class derived from <xref:Microsoft.AspNetCore.Mvc.ActionResult>), not a type like `string`.

Reviewer: avoid 1 line bullets
